docs: Documentation for MetricPipeline (#362)
kyma-project/telemetry-manager: docs/user/04-metrics.md

# Metrics

> >**Note**: The feature is not available yet. To understand the current progress, watch this [epic](https://github.com/kyma-project/kyma/issues/13079).

Reviewer: ```suggestion
> **NOTE:** The feature is not available yet. To understand the current progress, watch this [epic](https://github.com/kyma-project/kyma/issues/13079).
```

> Observability is all about exposing the internals of components in a distributed application and making that internal analyzable at a central place.

Reviewer: I don't understand this sentence.

> Observability is all about exposing the internals of components in a distributed application and making that internal analyzable at a central place.

Author: ```suggestion
Observability is all about exposing the internals of the components belonging to an distributed application and making that data analysable at a central place.
```

> While application logs and traces are usually request oriented, metrics are aggregated statistics exposed by a component to reflect the internal state. Typical statistics like the amount of processed requests, or the amount registered users can be very useful to introspect the current state and also the health of a component. Also, it allows to define proactive and reactive alerts in case of metrics are about to reach thresholds soon or in case they already passed thresholds.

Reviewer: ```suggestion
While application logs and traces are usually request-oriented, metrics are aggregated statistics exposed by a component to reflect the internal state. Typical statistics like the amount of processed requests, or the amount of registered users, can be very useful to introspect the current state and also the health of a component. Also, you can define proactive and reactive alerts if metrics are about to reach thresholds, or if they already passed thresholds.
```

> The goal of the Telemetry Module is to support you in collecting all relevant metrics of workload in a Kyma cluster and ship them to a backend for further analysis. Hereby, relevant Kyma modules like Istio or Serverless will contribute instantly and typical enrichment of the data is happening. Multiple [vendors for OTLP-based backends](https://opentelemetry.io/ecosystem/vendors/) are available.

Reviewer: ```suggestion
The goal of Kyma's Telemetry module is to support you in collecting all relevant metrics of a workload in a Kyma cluster and ship them to a backend for further analysis. Relevant Kyma modules like Istio or Serverless will contribute metrics instantly, and the Telemetry module enriches the data. You can choose among multiple [vendors for OTLP-based backends](https://opentelemetry.io/ecosystem/vendors/).
```

## Prerequisites

> A component from which you want to collect metrics data, needs to expose (or instrument) the metrics first. Typically, it instruments specific metrics for the used language runtime (like nodeJS) and custom metrics specific to the business logic. Also, the exposure can be in different formats, usually the pull-based prometheus format or the [push-based OTLP format](https://opentelemetry.io/docs/specs/otlp/).

Reviewer: ```suggestion
Before you can collect metrics data from a component, it must expose (or instrument) the metrics first. Typically, it instruments specific metrics for the used language runtime (like Node.js) and custom metrics specific to the business logic. Also, the exposure can be in different formats, like the pull-based Prometheus format or the [push-based OTLP format](https://opentelemetry.io/docs/specs/otlp/).
```

> To do the instrumentation you usually leverage an SDK, namely the [prometheus-client libraries](https://prometheus.io/docs/instrumenting/clientlibs/) or the [Open Telemetry SDKs](https://opentelemetry.io/docs/instrumentation/). Both libraries provide extensions to activate language specific auto-instrumentation like for nodeJS and an API to implement custom instrumentation.

Reviewer: ```suggestion
For the instrumentation, you usually use an SDK, namely the [Prometheus client libraries](https://prometheus.io/docs/instrumenting/clientlibs/) or the [Open Telemetry SDKs](https://opentelemetry.io/docs/instrumentation/). Both libraries provide extensions to activate language-specific auto-instrumentation like for Node.js, and an API to implement custom instrumentation.
```

## Architecture

> The Telemetry module provides an in-cluster central Deployment of an [OTel Collector](https://opentelemetry.io/docs/collector/) acting as a gateway. The gateway exposes endpoints for the [OTLP protocol](https://opentelemetry.io/docs/specs/otlp/) for GRPC and HTTP-based communication using the dedicated `telemetry-otlp-metrics` service, where all Kyma components and users' applications should send the metrics data to.

Reviewer: ```suggestion
In the Telemetry module, a central in-cluster Deployment of an [OTel Collector](https://opentelemetry.io/docs/collector/) acts as a gateway. The gateway exposes endpoints for the [OTLP protocol](https://opentelemetry.io/docs/specs/otlp/) for GRPC and HTTP-based communication using the dedicated `telemetry-otlp-metrics` service, to which all Kyma components and users' applications should send the metrics data.
```

> Optional, it provides a DaemonSet of an [OTel Collector](https://opentelemetry.io/docs/collector/) acting as an agent. That agent will scrape metrics of workload in the [prometheus pull-based format](https://prometheus.io/docs/instrumenting/exposition_formats) and can provide runtime specific metrics for workload.

Reviewer: ```suggestion
Optionally, the Telemetry module provides a DaemonSet of an [OTel Collector](https://opentelemetry.io/docs/collector/) acting as an agent. That agent scrapes metrics of workload in the [Prometheus pull-based format](https://prometheus.io/docs/instrumenting/exposition_formats) and can provide runtime specific metrics for workload.
```

> ![Architecture](./assets/metrics-arch.drawio.svg)

Reviewer: In the diagram, some connector descriptions have white background. To match the other ones, they need transparent background.
Also, "Prometheus" must be capitalized

Author: fixed

> 1. An application exposing metrics in OTLP, pushes metrics to the central metric gateway service 

Reviewer: ```suggestion
1. An application exposing metrics in OTLP, pushes metrics to the central metric gateway service.
```

> 2. An application exposing metrics in prometheus protocol, activates the agent to scrape the metrics via aan annotation based configuration

Reviewer: ```suggestion
2. An application exposing metrics in Prometheus protocol, activates the agent to scrape the metrics with an annotation-based configuration.
```

> 3. The agent can be activated to scrape metrics of each istio sidecar additionally

Reviewer: ```suggestion
3. Additionally, you can activate the agent can be activated to scrape metrics of each Istio.
```

> 4. The agent convertes and pushes all scraped metric data to the gateway in OTLP

Reviewer: ```suggestion
4. The agent converts and pushes all scraped metric data to the gateway in OTLP.
```

> 5. The gateway enrichs all received data with typical metadata of the source by communicating with the K8S APIServer. Furthermore, it will filter data according to the pipeline configuration.

Reviewer: ```suggestion
5. The gateway enriches all received data with typical metadata of the source by communicating with the Kubernetes APIServer. Furthermore, it filters data according to the pipeline configuration.
```

> 6. With the `MetricPipeline` resource, the metric gateway is configured with a target backend.

Reviewer: ```suggestion
6. The `MetricPipeline` resource specifies the target backend for the metric gateway.
```

> 1. The backend can run in-cluster.

Reviewer: ```suggestion
1. The backend can run within the cluster.
```

> 1. The backend can also run out-cluster, if authentication has been set up.

Reviewer: ```suggestion
1. If authentication has been set up, the backend can also run outside the cluster.
```

> 1. The metric data can be consumed using the backend system.

Reviewer: Why "can" ? Is this optional? What else do you do with the metric data?

Could we try:
```suggestion
1. The metric data is consumed using the backend system.
```

### Metric Gateway

> In a Kyma cluster, the Metric Gateway is the central component to which all components can send their individual metrics. The gateway collects, enriches, and dispatches the data to the configured backend. The gateway is based on the [OTel Collector](https://opentelemetry.io/docs/collector/) and comes with a [concept](https://opentelemetry.io/docs/collector/configuration/) of pipelines consisting of receivers, processors, and exporters, with which you can flexibly plug pipelines together. Kyma's MetricPipeline provides a hardened setup of an OTel Collector and also abstracts the underlying pipeline concept. Such abstraction has the following benefits:

Reviewer: Why is "metric gateway" capitalized here but nowhere else?

```suggestion
In a Kyma cluster, the metric gateway is the central component to which all components can send their individual metrics. The gateway collects, enriches, and dispatches the data to the configured backend. The gateway is based on the [OTel Collector](https://opentelemetry.io/docs/collector/) and comes with a concept of pipelines consisting of receivers, processors, and exporters, with which you can flexibly plug pipelines together (see [Configuration](https://opentelemetry.io/docs/collector/configuration/). Kyma's MetricPipeline provides a hardened setup of an OTel Collector and also abstracts the underlying pipeline concept. Such abstraction has the following benefits:
```

- Supportability - all features are tested and supported
- Migratability - smooth migration experiences when switching underlying technologies or architectures

> - Native Kubernetes support - API provided by Kyma allows for an easy integration with Secrets, for example, served by the [SAP BTP Service Operator](https://github.com/SAP/sap-btp-service-operator#readme). Telemetry Manager takes care of the full lifecycle.

Reviewer: ```suggestion
- Native Kubernetes support - API provided by Kyma supports an easy integration with Secrets, for example, served by the [SAP BTP Service Operator](https://github.com/SAP/sap-btp-service-operator#readme). Telemetry Manager takes care of the full lifecycle.
```

- Focus - the user doesn't need to understand underlying concepts

The downside is that only a limited set of features is available. If you want to avoid this downside, bring your own collector setup. The current feature set focuses on providing the full configurability of backends integrated by OTLP.

### Metric Agent

> If a MetricPipeline configures a feature in the `input.application` section, an additional DaemonSet gets deployed acting as an agent. The agent is as well based on an [OTel Collector](https://opentelemetry.io/docs/collector/) and is encompassing the collection and conversion of prometheus-based metrics. Hereby, the workload simply puts an typical `prometheus.io/scrape` annotation onn the pod or service specification and the agent will collect scraping it. The agent will push all data in OTLP to the central gateway.

Reviewer: ```suggestion
If a MetricPipeline configures a feature in the `input.application` section, an additional DaemonSet is deployed acting as an agent. The agent is also based on an [OTel Collector](https://opentelemetry.io/docs/collector/) and encompasses the collection and conversion of Prometheus-based metrics. Hereby, the workload simply puts a typical `prometheus.io/scrape` annotation on the specification of the Pod or service, and the agent collects it. The agent pushes all data in OTLP to the central gateway.
```

> ### Telemetry Manager

Reviewer: ```suggestion

### Telemetry Manager
```

The MetricPipeline resource is managed by Telemetry Manager, a typical Kubernetes [operator](https://kubernetes.io/docs/concepts/extend-kubernetes/operator/) responsible for managing the custom parts of the OTel Collector configuration.

![Manager resources](./assets/metrics-resources.drawio.svg)

> Telemetry Manager watches all MetricPipeline resources and related Secrets. Whenever the configuration changes, it validates the configuration and generates a new configuration for the gateway and agent, where for each a ConfigMap for the configuration is generated. Referenced Secrets are copied into one Secret that is mounted to the gateway as well.

Reviewer: ```suggestion
Telemetry Manager watches all MetricPipeline resources and related Secrets. Whenever the configuration changes, it validates the configuration and generates a new configuration for the gateway and agent, and for each a ConfigMap for the configuration is generated. Referenced Secrets are copied into one Secret that is mounted to the gateway as well.
```

> Furthermore, the manager takes care of the full lifecycle of the Gateway Deployment and the Agent DaemonSet itself. Only if there is a MetricPipeline defined, they are deployed. At anytime, you can opt out of using the feature by not specifying a MetricPipeline.

Reviewer: why are "Gateway" and "Agent" capitalized here?

## Setting up a MetricPipeline

In the following steps, you can see how to set up a typical MetricPipeline. Learn more about the available [parameters and attributes](resources/05-metricpipeline.md).

### Step 1. Create a MetricPipeline with an output
1. To ship metrics to a new OTLP output, create a resource file of the kind `MetricPipeline`:

   ```yaml
   apiVersion: telemetry.kyma-project.io/v1alpha1
   kind: MetricPipeline
   metadata:
     name: backend
   spec:
     output:
       otlp:
         endpoint:
           value: https://backend.example.com:4317
   ```

>    This configures the underlying OTel Collector of the gateway with a pipeline for metrics. The receiver of the pipeline will be of the OTLP type and be accessible using the `telemetry-otlp-metrics` service. As an exporter, an `otlp` or an `otlphttp` exporter is used, dependent on the configured protocol.

Reviewer: ```suggestion
   This configures the underlying OTel Collector of the gateway with a pipeline for metrics. The receiver of the pipeline will be of the OTLP type and be accessible using the `telemetry-otlp-metrics` service. As an exporter, an `otlp` or an `otlphttp` exporter is used, depending on the configured protocol.
```

2. To create the instance, apply the resource file in your cluster:
    ```bash
    kubectl apply -f path/to/my-metric-pipeline.yaml
    ```

3. Check that the status of the MetricPipeline in your cluster is `Ready`:
    ```bash
    kubectl get metricpipeline
    NAME              STATUS    AGE
    http-backend      Ready     44s
    ```

### Step 2. Switch the protocol to HTTP

> To use the HTTP protocol instead of the default GRPC, use the `protocol` attribute and ensure that the proper port is configured as part of the endpoint. Typically, port `4317` is used for GRPC and port `4318` for HTTP.

Reviewer: ```suggestion
To use the HTTP protocol instead of the default GRPC, use the `protocol` attribute and ensure that the correct port is configured as part of the endpoint. Typically, port `4317` is used for GRPC and port `4318` for HTTP.
```

```yaml
apiVersion: telemetry.kyma-project.io/v1alpha1
kind: MetricPipeline
metadata:
  name: backend
spec:
  output:
    otlp:
      protocol: http
      endpoint:
        value: https://backend.example.com:4318
```

### Step 3: Add authentication details

> To integrate with external systems, you must configure authentication details. At the moment, Basic Authentication and custom headers are supported.

Reviewer: ```suggestion
To integrate with external systems, you must configure authentication details. At the moment, Basic Authentication and custom headers are supported.

See the following code examples for mTLS, basic, and token-based custom authentication:
```

      otlp:
        endpoint:
          value: https://backend.example.com/otlp:4317
        tls:
          cert:
            value: |
              -----BEGIN CERTIFICATE-----
              ...
          key:
            value: |
              -----BEGIN RSA PRIVATE KEY-----
              ...
  ```
  </details>
  <details>
    <summary>Basic authentication</summary>

  ```yaml
  apiVersion: telemetry.kyma-project.io/v1alpha1
  kind: MetricPipeline
  metadata:
    name: backend
  spec:
    output:
      otlp:
        endpoint:
          value: https://backend.example.com/otlp:4317
        authentication:
          basic:
            user:
              value: myUser
            password:
              value: myPwd
  ```
  </details>
  <details>
    <summary>Token-based with custom headers</summary>

  ```yaml
  apiVersion: telemetry.kyma-project.io/v1alpha1
  kind: MetricPipeline
  metadata:
    name: backend
  spec:
    output:
      otlp:
        endpoint:
          value: https://backend.example.com/otlp:4317
        headers:
          - name: Authorization
            value: "Bearer myToken"
  ```
  </details>
</div>

### Step 4: Add authentication details from Secrets

> Integrations into external systems usually require authentication details dealing with sensitive data. To handle that data properly in Secrets, MetricsPipeline supports the reference of Secrets.

Reviewer: ```suggestion
Integrations into external systems usually need authentication details dealing with sensitive data. To handle that data properly in Secrets, MetricsPipeline supports the reference of Secrets.
```

  kind: MetricPipeline
  metadata:
    name: backend
  spec:
    output:
      otlp:
        endpoint:
          valueFrom:
              secretKeyRef:
                  name: backend
                  namespace: default
                  key: endpoint
        authentication:
          basic:
            user:
              valueFrom:
                secretKeyRef:
                  name: backend
                  namespace: default
                  key: user
            password:
              valueFrom:
                secretKeyRef:
                  name: backend
                  namespace: default
                  key: password
  ```
</details>
  <details>
    <summary>Token-based with custom headers</summary>

  ```yaml
  apiVersion: telemetry.kyma-project.io/v1alpha1
  kind: MetricPipeline
  metadata:
    name: backend
  spec:
    output:
      otlp:
        endpoint:
          value: https://backend.example.com:4317
        headers:
          - name: Authorization
            valueFrom:
              secretKeyRef:
                  name: backend
                  namespace: default
                  key: token 
  ```
  </details>
</div>

> The related Secret must have the referenced name and needs to be located in the referenced Namespace, and contain the mapped key as in the following example:

Reviewer: ```suggestion
The related Secret must have the referenced name, must be located in the referenced Namespace, and must contain the mapped key as in the following example:
```

```yaml
kind: Secret
apiVersion: v1
metadata:
  name: backend
  namespace: default
stringData:
  endpoint: https://backend.example.com:4317
  user: myUser
  password: XXX
  token: Bearer YYY
```

### Step 5: Rotate the Secret

Telemetry Manager continuously watches the Secret referenced with the **secretKeyRef** construct. You can update the Secret’s values, and Telemetry Manager detects the changes and applies the new Secret to the setup.
If you use a Secret owned by the [SAP BTP Service Operator](https://github.com/SAP/sap-btp-service-operator), you can configure an automated rotation using a `credentialsRotationPolicy` with a specific `rotationFrequency` and don’t have to intervene manually.

### Step 6: Activate Prometheus-based metrics

> To enable collection of prometheus based metrics define a MetricPipeline having prometheus enabled as input:

Reviewer: ```suggestion

> **NOTE:** For the following approach, you must have instrumented your application using a library like the [Prometheus client library](https://prometheus.io/docs/instrumenting/clientlibs/), with a port in your workload exposed serving as a Prometheus metrics endpoint.

To enable collection of Prometheus-based metrics, define a MetricPipeline that has Prometheus enabled as input:
```

```yaml
apiVersion: telemetry.kyma-project.io/v1alpha1
kind: MetricPipeline
metadata:
  name: backend
spec:
  input:
    application:
      prometheus:
        enabled: true
  output:
    otlp:
      endpoint:
        value: https://backend.example.com:4317
```

> This approach assumes that you instrumented your application using a library like the [prometheus client library](https://prometheus.io/docs/instrumenting/clientlibs/), having a port in your workload exposed serving a typical Prometheus metrics endpoint.

Reviewer: This seems like it would be good to know before the instruction

```suggestion
```

The agent is configured with a generic scrape configuration, which uses annotations to specify the endpoints to scrape in the cluster. 

> Having the annotations in place is everything you need for metrics ingestion to start automatically.

Reviewer: ```suggestion
You only need to have the annotations in place for metrics ingestion to start automatically.
```

Put the following annotations either to a service that resolves your metrics port, or directly to the pod:

```yaml
prometheus.io/scrape: "true"   # mandatory to enable automatic scraping
prometheus.io/scheme: https    # optional, default is "http" if no Istio sidecar is used. When using a sidecar (Pod has label `security.istio.io/tlsMode=istio`), the default is "https". Use "https" to scrape workloads using Istio client certificates.
prometheus.io/port: "1234"     # optional, configure the port under which the metrics are exposed
prometheus.io/path: /myMetrics # optional, configure the path under which the metrics are exposed
```

> **NOTE:** The agent can scrape endpoints even if the workload uses Istio and accepts only mTLS communication.

### Step 7: Activate runtime metrics

> To enable collection of runtime metrics for your pods, define a MetricPipeline having runtime enabled as input:

Reviewer: ```suggestion
To enable collection of runtime metrics for your Pods, define a MetricPipeline with runtime enabled as input:
```

```yaml
apiVersion: telemetry.kyma-project.io/v1alpha1
kind: MetricPipeline
metadata:
name: backend
spec:
  input:
    application:
      runtime:
        enabled: true
  output:
    otlp:
      endpoint:
        value: https://backend.example.com:4317
```

### Step 8: Activate Istio metrics

> To enable collection of istio metrics for your pods, define a MetricPipeline having istio enabled as input:

Reviewer: ```suggestion
To enable collection of Istio metrics for your Pods, define a MetricPipeline with Istio enabled as input:
```

Is step 8 additional to step 7, or are they alternatives to each other?

Author: can be mixed as well, they are not exclusive to each other

```yaml
apiVersion: telemetry.kyma-project.io/v1alpha1
kind: MetricPipeline
metadata:
name: backend
spec:
  input:
    application:
      istio:
        enabled: true
  output:
    otlp:
      endpoint:
        value: https://backend.example.com:4317
```

> 

Reviewer: Looks like the instructions are finished with step 8.

Can we provide a "Result" section that describes how the user knows they followed the instructions successfully?

Author: ok, I changed the steps do be aggregatable and added a result section which I mainly removed from the first step, as in the first step it is too early to try it out.
I applied changes to all three documents.

## Limitations

> The metric gateway setup is designed using the following assumptions:

Reviewer: ```suggestion
The metric gateway setup is based on the following assumptions:
```

- The collector has no autoscaling options yet and has a limited resource setup of 1 CPU and 1 GiB memory.
- Batching is enabled, and a batch will contain up to 512 metrics/batch.

> - An unavailability of a destination must be survived for 5 minutes without direct loss of metric data.

Reviewer: ```suggestion
- A destination can be unavailable for up to 5 minutes without direct loss of metric data.
```

What does "direct" mean? Can data be lost indirectly?

- An average metric consists of 40 attributes with 64 character length.

This leads to the following limitations:
### Throughput

> The maximum throughput is 4200 metric/sec ~= 15.000.000 metrics/hour. If more data must be ingested, it can result in a refusal of more data.

Reviewer: ```suggestion
The maximum throughput is 4200 metric/sec ~= 15.000.000 metrics/hour. If more data must be ingested, it can be refused.
```

### Unavailability of output
For up to 5 minutes, a retry for data is attempted when the destination is unavailable. After that, data is dropped.

### No guaranteed delivery
The used buffers are volatile. If the gateway or agent instances crash, metric data can be lost.

### Multiple MetricPipeline support

> Up to three MetricPipeline resources at a time are supported at the moment.

Reviewer: ```suggestion
Up to three MetricPipeline resources at a time are supported.
```

Once we support more pipelines, we announce it as a feature and describe it here. No need to mention it yet.
